encrypt.go

```diff
@@ -4,31 +4,133 @@ import (
 	"crypto/aes"
 	"crypto/cipher"
 	"crypto/rand"
-	"crypto/sha256"
+	"crypto/sha512"
+	"encoding/hex"
 	"errors"
+	"fmt"
 	"io"
 	"io/ioutil"
+	"strings"
 
 	"golang.org/x/crypto/pbkdf2"
 )
 
 const (
-	keylen  = 32
-	saltlen = 8
+	// if these settings take too long on your server or workstation feel free to modify
+	// however please keep in mind these are the settings that Temporal uses.
+	// thus if you want to decrypt a file which was encrypted by our Temporal node, you must
+	// ensure that the settings match as foillows: keylen = 32, saltlen = 32, nonceSize = 24
+	keylen    = 32
+	saltlen   = 32
+	nonceSize = 24
+)
+
+// Protocol is used to configure encryption/decryption methods
+type Protocol string
+
+var (
+	// GCM allows for usage of AES256-GCM encryption/decryption
+	GCM Protocol = "AES256-GCM"
+	// CFB allows for usage of AES256-CFB encryption/decryption
+	CFB Protocol = "AES256-CFB"
 )
 
 // EncryptManager handles file encryption and decryption
 type EncryptManager struct {
-	passphrase []byte
+	passphrase       []byte
+	gcmDecryptParams *GCMDecryptParams
+	protocol         Protocol
+}
+
+// GCMDecryptParams is used to configure decryption for AES256-GCM
+type GCMDecryptParams struct {
+	CipherKey string
+	Nonce     string
 }
 
 // NewEncryptManager creates a new EncryptManager
-func NewEncryptManager(passphrase string) *EncryptManager {
-	return &EncryptManager{[]byte(passphrase)}
+// If instantiating to use CFB encryption/decryption, gcmDecrypt params can be null
+// If instantiating to use GCM encryption, gcmDecrypt params can be null
+// If instantiating to use GCM decryption, gcmDecrypt params must not be null
+func NewEncryptManager(passphrase string, protocol Protocol, gcmDecryptParams *GCMDecryptParams) *EncryptManager {
```

Comment:
I think instead of hardcoding this param here, have a function `EncryptManager.WithGCM(GCMDecryptParams)`, then usage would be:

```go
b, err := NewEncryptManager("hello", GCM).
    WithGCM(params).
    Decrypt(data)
```

Reply:
to avoid overloading the constructor with parameters

Reply:
WOW now that I think of it this would be pretty clean:

```go
var params *GCMDecryptParams
b, err := crypto.NewEncryptManager("hello").
    WithGCM(params).
    Encrypt(data)

// print details for decryption
fmt.Printf("%v", params)
```

and:

```go
// set details for decryption
var params = &GCMDecryptParams{ /* ... */ }
b, err := crypto.NewEncryptManager("hello").
    WithGCM(params).
    Decrypt(data)
```

Reply:
so `withGCM` will set [EncryptManager::protocol](https://github.com/RTradeLtd/crypto/pull/2/files#diff-3c7b653a50150ab2dd32024a5aa2d2b6R42)

Reply:
SLICK, so we can have two functions `WithGCM` and `WithCFB` which would take in the respective arguments, and return an `EncryptManager` to be used with the `Decrypt` functino?

Reply:
yeah i think so

could forgo `WithCFB` and leave it as the default, since it requires no additional arguments

Reply:
I think it makes sense to use WithCFB, since during instantiation we can omit the protocol, and have `WithCFB` set protocol to CFB, etc...

Reply:
yeah but it might be useful to have a default, right? given none of the `With___` functions are called

Reply:
Good point